Add iOS Phone Number Auth

pranjalsatija opened the pull request:
iOS phone number auth, as mentioned in #329. Looking at this vs. the Android implementation in the existing PR, I think we could benefit from agreeing on some sort of standard API in terms of the names of the methods and parameters we want to expose between native code and Flutter. I tried to piece this together using what I saw in @alibitek's implementation, but I'm not sure if it's 100%, so maybe we could talk about that?

pranjalsatija commented:
Signed it.

mravn-google commented on packages/firebase_auth/ios/Classes/FirebaseAuthPlugin.m:
```diff
@@ -113,6 +131,27 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
+                                              completion:^(NSString *verificationID, NSError *error) {
+                                                  if (error == nil) {
+                                                      self.phoneAuthVerificationID = verificationID;
+                                                      _phoneAuthEventSink(@"CODE_SENT");
```
May be `nil`, in which case we should not attempt to emit the event.

mravn-google commented on packages/firebase_auth/ios/Classes/FirebaseAuthPlugin.m:
```diff
@@ -30,7 +30,10 @@ - (FlutterError *)flutterError {
-@property(nonatomic, retain) FlutterMethodChannel *channel;
+@property(nonatomic, retain) FlutterMethodChannel *methodChannel;
+@property(nonatomic, retain) FlutterEventChannel *phoneAuthEventChannel;
+@property(nonatomic, copy) FlutterEventSink phoneAuthEventSink;
```
Event streams seem a bit heavy for just emitting a single event. If there are no other uses for the stream, maybe just use a method call in the iOS to Dart direction?

pranjalsatija replied:
I also think the event stream setup is overkill for this, but I did it because [the Android implementation](ae74f4c4532ff796ebfb8b54d2adc8079c49202a) used it. Android's phone number auth seems to support a lot more features than iOS' implementation does. I just looked through that linked implementation and tried to piece something together. I think it would serve us well to define what we want the API to look like so there's no confusion about things like method names, parameter information, and communication between native code and Flutter.

mravn-google commented:
@collinjackson Assigning you since this seems closely related to https://github.com/flutter/plugins/pull/329. Feel free to reassign to me, if that is wrong.

pranjalsatija commented:
@mravn-google Thanks for the feedback!

kroikie commented:
Hi @pranjalsatija thank you for working on this feature with @alibitek.

Given that the options of the phone auth are different for each side of this plugin let us work with the interface used by the [Unity plugin](https://firebase.google.com/docs/auth/unity/phone-auth#send-a-verification-code-to-the-users-phone).

The Dart interface would look like:

```
verifyPhoneNumber(phoneNumber, timeout,
    verificationCompleted: (credential) => {},
    verificationFailed: (error) => {},
    codeSent: (id, token) => {},
    codeAutoRetrievalTimeout: (id) => {}
)
```

pranjalsatija commented:
@kroikie I just committed the Dart interface for phone number auth. Could you take a look? It's my first Dart / Flutter PR, so there might need to be some changes. If everything looks good, let me know and I'll get started with the iOS implementation.

kroikie commented:
@pranjalsatija `verifyPhoneNumber` looks good, there will need to be an implementation of `getCredential` so the dev can pass in a verificationID and SMS code.

alibitek commented:
@pranjalsatija You have some code formatting problems. You can correct them by running:
```
./script/plugin_tools.sh format
./script/plugin_tools.sh analyze
./script/plugin_tools.sh test
```
or
```
pub global activate flutter_plugin_tools
pub global run flutter_plugin_tools format --plugins=firebase_auth
pub global run flutter_plugin_tools analyze --plugins=firebase_auth
pub global run flutter_plugin_tools test --plugins=firebase_auth
```
from the root of the Flutter plugins repository, as mentioned in [CONTRIBUTING.md](https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#contributing-code)

kroikie commented:
@pranjalsatija thanks for your submission, it is very close some small adjustments and I think we can move fwd with this feature:

Here is the updated signature proposed:

```
class AuthException implements Exception {
  final String code;
  final String message;
  const AuthException(this.code, this.message);
}

typedef void PhoneVerificationCompleted(FirebaseUser firebaseUser);
typedef void PhoneVerificationFailed(AuthException error);
typedef void PhoneCodeSent(String verificationId);
typedef void PhoneCodeAutoRetrievalTimeout(String verificationId);

Future<void> verifyPhoneNumber({
    @required String phoneNumber,
    @required Duration timeout,
    bool resend = false,
    @required PhoneVerificationCompleted verificationCompleted,
    @required PhoneVerificationFailed verificationFailed,
    @required PhoneCodeSent codeSent,
    @required PhoneCodeAutoRetrievalTimeout codeAutoRetrievalTimeout,
  }) async {
  ...
}
```

matejthetree commented:
What is the current state of the phone auth on both iOS and Android? Is there some progress eta?

pranjalsatija commented:
@kroikie Just updated the Dart signatures for that on [my fork](https://github.com/pranjalsatija/plugins/commit/294adb5082075cd3d922a7384aeb703dea2bf42b). One thing to point out is that we lose access to the resend token with this existing API. Are we going to make the resend token only visible to the platform APIs internally?

pranjalsatija commented:
@kroikie Any idea why the CI is failing? When I run the scripts on my computer, I get this:

```
The following packages have analyzer errors (see above):
 * firebase_database
 * cloud_firestore
 * video_player
 * firebase_messaging
 * google_mobile_maps
 * firebase_analytics
 * image_picker
 * firebase_storage
 * shared_preferences
```

And all the errors in the packages listed are violations of the `prefer_const_constructors` lint rule.

kroikie commented:
Hi @pranjalsatija no sure why CI is failing, looking into it now and will update here.

pranjalsatija commented:
Sounds good. Should I open a PR to fix those const constructor warnings as well?

pranjalsatija commented:
@kroikie I rebased my own code and squashed my commits into aea1bc4042db444369ce06a542cd5e70f11e176d. I also rebased `flutter/plugins/master` onto my fork.

pranjalsatija commented:
@kroikie OK, it looks like `plugin_tools.sh format` is failing. I can't actually run it on my computer though because I get a lot of errors about files not being found, but they're present. Any advice on how to fix that?

kroikie commented:
Ensure that `pub` and `clang-format` are on your path. Then try running:
`pub global activate flutter_plugin_tools && pub global run flutter_plugin_tools format`

pranjalsatija commented:
I already did that. It successfully analyzes all the `.dart` and `.java` files, but crashes when it gets to `.h` and `.m` files:

```
Formatting all .m and .h files...
Unhandled exception:
ProcessException: No such file or directory
  Command: clang-format -i --style=Google /Users/pranjalsatija/Documents/Programming/Open Source/plugins/packages/path_provider/example/ios/Runner/AppDelegate.h
...
```

It continues on and lists *many* files, all of which are Objective C source that's present on my machine. It also looks like those sources are from a whole bunch of different packages, not just the one I've been working on.

kroikie commented:
This usually indicates that `clang-format` is not on your path or cannot be reached by the script. Make sure that it is available in your terminal then the script should complete.

pranjalsatija commented:
Got it. I had it in my path before, but it stopped working today. I'm pretty sure it's because of Xcode 10, so I'm going to try this again with Xcode 9.

kroikie commented:
Hi @pranjalsatija I have combined both the Android and iOS version of this feature into a [single PR](https://github.com/flutter/plugins/pull/606). I will close this PR and welcome your feedback in the combined PR.
